Products/ZenHub/invalidationmanager.py

```diff
@@ -0,0 +1,281 @@
+            if self._paused():
+                return
+
+            oids = self._poll_invalidations()
+            if not oids:
+                log.debug('no invalidations found: oids=%s' % oids)
+                return
+
+            for oid in oids:
+                yield self.invalidation_pipeline.run(oid)
+
+            self.log.debug('Processed %s raw invalidations', len(oids))
+
+        except Exception:
+            log.exception('error in process_invalidations')
+        finally:
+            self.totalEvents += 1
+            self.totalTime += time() - now
+            log.debug('end process_invalidations')
+
+    @inlineCallbacks
+    def _syncdb(self):
+        try:
+            self.log.debug("[processQueue] syncing....")
+            yield self.__syncdb()
+            self.log.debug("[processQueue] synced")
+        except Exception as err:
+            self.log.warn("Unable to poll invalidations, will try again.")
+
+    def _paused(self):
+        if not self._currently_paused:
+            if self.__dmd.pauseHubNotifications:
+                self._currently_paused = True
+                log.info('notifications have been paused')
+                self._send_event(self._invalidation_paused_event)
+                return True
+            else:
+                return False
+
+        else:
+            if self.__dmd.pauseHubNotifications:
+                log.debug('notifications are paused')
+                return True
+            else:
+                self._currently_paused = False
+                log.info('notifications unpaused')
+                self._send_event(self._invalidation_unpaused_event)
+                return False
+
+    def _poll_invalidations(self):
+        '''pull a list of invalidated object oids from the database
+        '''
+        try:
+            log.debug('poll invalidations from dmd.storage')
+            return self.__poll_invalidations()
+        except Exception:
+            log.exception('error in _poll_invalidations')
+
+    @inlineCallbacks
+    def _send_event(self, event):
+        yield self.__send_event(event)
+
+
+class InvalidationPipeline(object):
+
+    def __init__(self, app, filters, processor):
+        self.__pipeline = oid_to_obj(
+            app, processor,
+            filter_obj(
+                filters,
+                transform_obj(
+                    processor
+                )
+            )
+        )
+
+    def run(self, invalidations):
+        self.__pipeline.send(invalidations)
+
+
+def coroutine(func):
+    """Decorator for initializing a generator as a coroutine.
+    """
+    @wraps(func)
+    def start(*args, **kw):
+        coro = func(*args, **kw)
+        coro.next()
+        return coro
+    return start
+
+
+@coroutine
+def iterate(target):
+    """iterates over the iterable, sending each produced item to the target.
+    """
+    while True:
+        iterable = (yield)
+        for item in iterable:
+            target.send(item)
```

Comment:
The `iterate` coroutine isn't used and so can be removed.  It's essentially dead code.

Reply:
it has been removed